keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -0,0 +1,1256 @@
+changing the invalid fields during an update.
+
+##### Unit tests
+
+<!--
+In principle every added code should have complete unit test coverage, so providing
+the exact set of tests will not bring additional value.
+However, if complete unit test coverage is not possible, explain the reason of it
+together with explanation why this is acceptable.
+-->
+
+<!--
+Additionally, for Alpha try to enumerate the core package you will be touching
+to implement this enhancement and provide the current unit coverage for those
+in the form of:
+- <package>: <date> - <current test coverage>
+The data can be easily read from:
+https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit
+
+This can inform certain test coverage improvements that we want to do before
+extending the production code to implement this enhancement.
+-->
+
+- `apiextensions-apiserver/pkg/apiserver/schema/cel`: `<date>` - `<test coverage>`
+- `apiextensions-apiserver/pkg/apiserver/schema/listtype`: `<date>` - `<test coverage>`
+- `apiextensions-apiserver/pkg/apiserver/schema/objectmeta`: `<date>` - `<test coverage>`
+
+##### Integration tests
+
+<!--
+Integration tests are contained in k8s.io/kubernetes/test/integration.
+Integration tests allow control of the configuration parameters used to start the binaries under test.
+This is different from e2e tests which do not allow configuration of parameters.
+Doing this allows testing non-default options and multiple different and potentially conflicting command line options.
+-->
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+-->
+
+1. Show normal creation path is unaffected
+2. Tests that ratchet different classes of validations:
+    - JSON-Schema fields (minItems, dependencies, etc.)
+    - CRD-Validation Rules
+    - ListSets/Maps
+3. Tests that certain classes of valdiations are not ratcheted:
+    - Nested Validations (not, oneOf, allOf, anyOf)
+    - Transition Rules
+
+##### e2e tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+
+We expect no non-infra related flakes in the last month as a GA graduation criteria.
+-->
+
+It would be nice to have a e2e test that tests ratcheting over a k8s upgrade
+which tightens a validation.
+
+- <test>: <link to test coverage>
+
+### Graduation Criteria
+
+
+#### Alpha
+- Feature Flag added
+- Integration Tests implemented and enabled
+- Schema-Based Ratcheting Implemented
+- List-Type Ratcheting Implemented
+- ObjectMeta/TypeMeta object-level ratcheting
+
+#### Beta
```

Comment:
Plus: 

- decide about opt-in or by-default ratcheting

Reply:
Do you mean at the feature gate level or the CRD-level? For CRDs a brief discussion is included in the alternatives section.

Reply:
CRD level. `spec.ratcheting: true` or `x-kubernetes-ratcheting: true`.

Reply:
re: this discussion https://github.com/kubernetes/enhancements/pull/4013#discussion_r1228502082

We are sticking with on-by-default since one of the main goals of the KEP was to allow us to change formats, etc. on Kubernetes' side without breaking workflows. Opt-in for CRDs would not have this property.

Reply:
devil's advocate: if that's the argument, we could just add manual ratcheting for formats.

Reply:
Counter argument to that could be that ratcheting a few fields when users did not opt into ratcheting is surprising UX, and increases maintenance burden on K8s devs not just through the code complexity but knowing when updating any form of validation to now ratchet that field. Added to doc

Reply:
I expected on-by-default ratcheting